components/scheduler/executor_executors.go

```diff
@@ -0,0 +1,493 @@
+	request := &historyservice.TerminateWorkflowExecutionRequest{
+		NamespaceId: scheduler.NamespaceId,
+		TerminateRequest: &workflowservice.TerminateWorkflowExecutionRequest{
+			Namespace:           scheduler.Namespace,
+			WorkflowExecution:   &commonpb.WorkflowExecution{WorkflowId: target.WorkflowId},
+			Reason:              "terminated by schedule overlap policy",
+			Identity:            scheduler.identity(),
+			FirstExecutionRunId: target.RunId,
+		},
+	}
+	ctx, cancelFunc := context.WithTimeout(context.Background(), e.Config.ServiceCallTimeout())
+	defer cancelFunc()
+	_, err := e.HistoryClient.TerminateWorkflowExecution(ctx, request)
+	return translateError(err, "TerminateWorkflowExecution")
+}
+
+func (e executorTaskExecutor) cancelWorkflow(scheduler Scheduler, target *commonpb.WorkflowExecution) error {
+	request := &historyservice.RequestCancelWorkflowExecutionRequest{
+		NamespaceId: scheduler.NamespaceId,
+		CancelRequest: &workflowservice.RequestCancelWorkflowExecutionRequest{
+			Namespace:           scheduler.Namespace,
+			WorkflowExecution:   &commonpb.WorkflowExecution{WorkflowId: target.WorkflowId},
+			Reason:              "cancelled by schedule overlap policy",
+			Identity:            scheduler.identity(),
+			FirstExecutionRunId: target.RunId,
+		},
+	}
+	ctx, cancelFunc := context.WithTimeout(context.Background(), e.Config.ServiceCallTimeout())
+	defer cancelFunc()
+	_, err := e.HistoryClient.RequestCancelWorkflowExecution(ctx, request)
+	return translateError(err, "CancelWorkflowExecution")
+}
+
+// getRateLimiterPermission returns a delay for which the caller should wait
+// before proceeding. If an error is returned, execution should not proceed, and
+// reservation should be retried.
+func (e executorTaskExecutor) getRateLimiterPermission() (delay time.Duration, err error) {
+	// TODO - per-namespace rate limiting. TaskExecutor is more broad than a single namespace.
+	return
+}
+
+// recordAction updates the SchedulerInfo metadata block with StartWorkflowExecution's result.
+func (e executorTaskExecutor) recordAction(s Scheduler, result *schedulepb.ScheduleActionResult) {
+	tweakables := e.Config.Tweakables(s.Namespace)
+	s.Info.ActionCount++
+	s.Info.RecentActions = util.SliceTail(append(s.Info.RecentActions, result), tweakables.RecentActionCount)
+	if result.StartWorkflowResult != nil {
+		s.Info.RunningWorkflows = append(s.Info.RunningWorkflows, result.StartWorkflowResult)
+	}
+}
+
+func (e executorTaskExecutor) loadExecutor(node *hsm.Node) (Executor, error) {
+	prevExecutor, err := hsm.MachineData[Executor](node)
+	if err != nil {
+		return Executor{}, err
+	}
+
+	return Executor{
+		ExecutorInternal: prevExecutor.ExecutorInternal,
+	}, nil
+}
+
+// translateError converts a dependent service error into an application error.
+// Errors are classified between retryable and non-retryable.
+func translateError(err error, msgPrefix string) error {
```

Comment:
Not sure I understand why you need this translation here. Why can't you use the original error? What's the value in wrapping with ApplicationError? That's an SDK concept and isn't relevant here.

Reply:
:+1: temporal.ApplicationError is an sdk thing, don't use it (or any error wrapping) here